File: sdk/python/vcr_kit/verifier.py
"""Offline verification helpers for consent receipts."""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, Optional

try:  # pragma: no cover - optional dependency
    from nacl.signing import VerifyKey
except ImportError:  # pragma: no cover
    VerifyKey = None  # type: ignore

from .dids import InMemoryDidResolver, public_key_from_multibase
from .revocation import JsonRevocationRegistry


@dataclass
class VerificationResult:
    verified: bool
    reason: Optional[str] = None


class ConsentVerifier:
    """Verify consent receipts offline."""

    def __init__(
        self,
        resolver: InMemoryDidResolver,
        revocations: JsonRevocationRegistry | None = None,
    ) -> None:
        self.resolver = resolver
        self.revocations = revocations

    async def verify(
        self,
        credential: Dict[str, Any],
        *,
        at_time: datetime | None = None,
    ) -> VerificationResult:
        proof = credential.get("proof")
        if not proof:
            return VerificationResult(False, "Missing proof")

        unsigned = dict(credential)
        unsigned.pop("proof", None)
        payload = _canonicalize(unsigned)

        if VerifyKey is None:
            return VerificationResult(False, "PyNaCl is required for signature verification")

        try:
            doc = await self.resolver.resolve(str(credential["issuer"]))
        except Exception as exc:  # pragma: no cover - passthrough
            return VerificationResult(False, str(exc))

        verification_method_id = proof.get("verificationMethod")
        method = next(
            (vm for vm in doc.verificationMethod if vm.id == verification_method_id),
            None,
        )
        if not method:
            return VerificationResult(False, "Verification method missing from DID document")

        public_key = public_key_from_multibase(method.publicKeyMultibase)
        signature = _b64url_decode(proof.get("proofValue", ""))

        try:
            VerifyKey(public_key).verify(payload, signature)
        except Exception:  # pragma: no cover - signature failure
            return VerificationResult(False, "Signature verification failed")

        now = at_time or datetime.now(timezone.utc)
        expiry = credential.get("expirationDate")
        if expiry:
            exp_raw = str(expiry)
            if exp_raw.endswith("Z"):
                exp_raw = exp_raw[:-1] + "+00:00"
            exp_time = datetime.fromisoformat(exp_raw)
            if exp_time.tzinfo is None:
                exp_time = exp_time.replace(tzinfo=timezone.utc)
            if now > exp_time:
                return VerificationResult(False, "Credential expired")

        if self.revocations:
            revoked = await self.revocations.is_revoked(str(credential["id"]))
            if revoked:
                return VerificationResult(False, "Credential revoked")

        return VerificationResult(True)


def _canonicalize(value: Dict[str, Any]) -> bytes:
    return json.dumps(value, separators=(",", ":"), sort_keys=True).encode("utf-8")


def _b64url_decode(value: str) -> bytes:
    padded = value + "=" * ((4 - len(value) % 4) % 4)
    return base64.urlsafe_b64decode(padded.encode("ascii"))

File: packages/vcr-kit/src/verifier.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyConsentReceipt = verifyConsentReceipt;
exports.assertVerified = assertVerified;
const canonicalize_1 = __importDefault(require("canonicalize"));
const luxon_1 = require("luxon");
const crypto_js_1 = require("./crypto.js");
const dids_js_1 = require("./dids.js");
async function verifyConsentReceipt(credential, options) {
    if (!credential.proof) {
        return { verified: false, reason: 'Missing proof' };
    }
    const { proof } = credential;
    const unsigned = { ...credential };
    delete unsigned.proof;
    const serialized = (0, canonicalize_1.default)(unsigned);
    if (!serialized) {
        return { verified: false, reason: 'Unable to canonicalize credential' };
    }
    let doc;
    try {
        doc = await options.resolver.resolve(credential.issuer);
    }
    catch (error) {
        return { verified: false, reason: error.message };
    }
    const verificationMethod = doc.verificationMethod.find((vm) => vm.id === proof.verificationMethod);
    if (!verificationMethod) {
        return { verified: false, reason: 'Verification method not found in DID document' };
    }
    const publicKey = (0, dids_js_1.publicKeyFromMultibase)(verificationMethod.publicKeyMultibase);
    const signature = (0, crypto_js_1.base64UrlDecode)(proof.proofValue);
    const message = Buffer.from(serialized);
    const valid = await (0, dids_js_1.verifyEd25519Signature)(publicKey, message, signature);
    if (!valid) {
        return { verified: false, reason: 'Signature verification failed' };
    }
    const now = resolveVerificationTime(options.atTime);
    const expiry = credential.expirationDate
        ? luxon_1.DateTime.fromISO(credential.expirationDate, { zone: 'utc' })
        : undefined;
    if (expiry && now > expiry) {
        return { verified: false, reason: 'Credential expired' };
    }
    if (options.revocationRegistry) {
        const revoked = await options.revocationRegistry.isRevoked(credential.id);
        if (revoked) {
            return { verified: false, reason: 'Credential revoked' };
        }
    }
    return { verified: true };
}
function assertVerified(result) {
    if (!result.verified) {
        throw new Error(result.reason ?? 'Verification failed');
    }
}
function resolveVerificationTime(atTime) {
    if (!atTime) {
        return luxon_1.DateTime.utc();
    }
    if (atTime instanceof Date) {
        return luxon_1.DateTime.fromJSDate(atTime).toUTC();
    }
    if (typeof atTime === 'string') {
        return luxon_1.DateTime.fromISO(atTime, { zone: 'utc' });
    }
    return atTime.toUTC();
}
